upstream/controllers.go

```diff
@@ -0,0 +1,103 @@
+package upstream
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/flanksource/commons/logger"
+	"github.com/flanksource/incident-commander/api"
+	"github.com/flanksource/incident-commander/db"
+	"github.com/labstack/echo/v4"
+	"gorm.io/gorm/clause"
+)
+
+func UpstreamPushesCtrl(c echo.Context) error {
+	var req api.PushData
+	err := json.NewDecoder(c.Request().Body).Decode(&req)
+	if err != nil {
+		return c.JSON(http.StatusBadRequest, api.HTTPErrorMessage{
+			Error:   err.Error(),
+			Message: "invalid json request",
+		})
+	}
+
+	if err := insertUpstreamMsg(c.Request().Context(), &req); err != nil {
+		return c.JSON(http.StatusInternalServerError, api.HTTPErrorMessage{
+			Error:   err.Error(),
+			Message: "something went wrong",
+		})
+	}
+
+	logger.Debugf("Checked at %v", req.CheckedAt)
+	return nil
+}
+
+func insertUpstreamMsg(ctx context.Context, req *api.PushData) error {
+	if len(req.Canaries) > 0 {
+		if err := db.Gorm.Clauses(clause.OnConflict{UpdateAll: true}).Create(req.Canaries).Error; err != nil {
```

Comment:
If processing one of this fails, none of the other fields would be processed. I don't think thats a desirable flow

Reply:
At this stage, we'll only have one item.